cmd/gsm-secret-sync/main.go

```diff
@@ -0,0 +1,190 @@
+}
+
+type ResourceManagerClient interface {
+	SetIamPolicy(ctx context.Context, req *iampb.SetIamPolicyRequest, opts ...gax.CallOption) (*iampb.Policy, error)
+	GetIamPolicy(ctx context.Context, req *iampb.GetIamPolicyRequest, opts ...gax.CallOption) (*iampb.Policy, error)
+}
+
+type IAMClient interface {
+	CreateServiceAccountKey(ctx context.Context, req *adminpb.CreateServiceAccountKeyRequest, opts ...gax.CallOption) (*adminpb.ServiceAccountKey, error)
+	CreateServiceAccount(ctx context.Context, req *adminpb.CreateServiceAccountRequest, opts ...gax.CallOption) (*adminpb.ServiceAccount, error)
+	DeleteServiceAccount(ctx context.Context, req *adminpb.DeleteServiceAccountRequest, opts ...gax.CallOption) error
+	ListServiceAccounts(ctx context.Context, req *adminpb.ListServiceAccountsRequest, opts ...gax.CallOption) *iamadmin.ServiceAccountIterator
+	ListServiceAccountKeys(ctx context.Context, req *adminpb.ListServiceAccountKeysRequest, opts ...gax.CallOption) (*adminpb.ListServiceAccountKeysResponse, error)
+	DeleteServiceAccountKey(ctx context.Context, req *adminpb.DeleteServiceAccountKeyRequest, opts ...gax.CallOption) error
+}
+
+func main() {
+	o := parseOptions()
+	if err := o.Validate(); err != nil {
+		logrus.WithError(err).Fatal("Failed to validate options")
+	}
+	if err := o.setupLogger(); err != nil {
+		logrus.WithError(err).Fatal("Failed to set up logging")
+	}
+
+	logrus.Info("Starting reconciliation")
+
+	desiredSAs, desiredSecrets, desiredIAMBindings, desiredCollections, err := gsm.GetDesiredState(o.configFile)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to parse configuration file")
+	}
+
+	ctx := context.Background()
+
+	projectsClient, err := resourcemanager.NewProjectsClient(ctx, option.WithQuotaProject(gsm.ProjectIdNumber))
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to create resource manager client")
+	}
+	defer projectsClient.Close()
+	policy, err := gsm.GetProjectIAMPolicy(ctx, projectsClient)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to get project IAM policy")
+	}
+
+	iamClient, err := iamadmin.NewIamClient(ctx, option.WithQuotaProject(gsm.ProjectIdNumber))
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to create IAM client")
+	}
+	actualSAs, err := gsm.GetUpdaterServiceAccounts(ctx, iamClient)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to get service accounts")
+	}
+
+	secretsClient, err := secretmanager.NewClient(ctx, option.WithQuotaProject(gsm.ProjectIdNumber))
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to create secrets client")
+	}
+	defer secretsClient.Close()
+	actualSecrets, err := gsm.GetSecrets(ctx, secretsClient)
+	if err != nil {
+		logrus.WithError(err).Fatal("Failed to get current secrets")
+	}
+
+	actions := gsm.ComputeDiff(desiredSAs, actualSAs, desiredSecrets, actualSecrets, desiredIAMBindings, policy, desiredCollections)
```

Comment:
I am wondering why GCS is not already doing the comparison. I guess you are checking the difference here to verify what exactly needs to change, but we can just do the call and change everything. Will we use more throttles?